### PR TITLE
Add warnings when client version is not compatible with the server

### PR DESCRIPTION
**Pull request by jkotilahti**
* The first commit adds a warning when initializing IQMClient connected to a server for which the reported compatible iqm-client version range (from `/info/client-libraries`) does not include the current version. The warning does not prevent using the client, since it is still possible that things work even if the version is officially incompatible. Example of how the warning looks like:

    `UserWarning: Your IQM Client version 21.0 was built for a different version of IQM Server. You might encounter issues. For the best experience, consider using a version of IQM Client that satisfies 20.0 <= iqm-client < 21.0.`

* The second commit improves the error message when an endpoint returns 404 and the client version is incompatible with the server. This can happen if an endpoint used by the current client version is removed in a new server version, or if the current client version adds support for using an endpoint that is not yet supported by the server (like happened in https://github.com/iqm-finland/qiskit-on-iqm/issues/121). Previously, it was not clear to users from the error message that the 404 was caused by version incompatibility, but now it would show for example this message:

    `requests.exceptions.HTTPError: https://test.qc.iqm.fi/cocos/api/v1/calibration/default/gates not found. This may be caused by the server version not supporting this endpoint. Your IQM Client version 21.0 was built for a different version of IQM Server. You might encounter issues. For the best experience, consider using a version of IQM Client that satisfies 18.0 <= iqm-client < 19.0.`

    For some endpoints, there was already special handling for 404 errors (for example it can mean a job with a specific id was not found), so for those ones I did not add this change.

### CODE REVIEW

**Review by freetonik — DISMISSED** [review]

File: src/iqm/iqm_client/iqm_client.py [hunk 1]
```diff
@@ -899,3 +911,39 @@ def _check_authentication_errors(result: requests.Response) -> None:
...
+            compatible_versions = versions_response.json()['iqm-client']
+            min_version = parse(compatible_versions['min'])
+            max_version = parse(compatible_versions['max'])
+            client_version = parse(version('iqm-client'))
+            if client_version < min_version or client_version >= max_version:
+                return (
+                    f'IQM Client version {client_version} is incompatible with the IQM server, '
+                    f'which requires {min_version} <= iqm-client < {max_version}. '
+                    f'Please use a compatible client version to make sure all features work correctly.'
+                )
```
Comment (manzanillo) [inline src/iqm/iqm_client/iqm_client.py:948]: As it is not automatically incompatible, I think, we should let it sound less like an error message but more liek a warning (as it is implementation-wise)
```suggestion
                return (
                    f'Your IQM Client version {client_version} was built for a different version of IQM Server. '
                    f'You might encounter issues. For the best experience, consider using a version '
                    f'of IQM Client that satisfies {min_version} <= iqm-client < {max_version}.'
                )
```

File: tests/test_iqm_client.py [hunk 2]
```diff
@@ -1006,3 +1009,63 @@ def test_get_dynamic_quantum_architecture_throws_json_decode_error_if_received_n
...
+            },
+        )
+    )
+    when(requests).get(f'{base_url}/api/v1/calibration/default/gates', ...).thenReturn(MockJsonResponse(404, {}))
+    with pytest.raises(
+        HTTPError,
+        match=re.escape(
+            f'IQM Client version {client_version} is incompatible with the IQM server, '
+            f'which requires {min_version} <= iqm-client < {max_version}'
+        ),
```
Comment (manzanillo) [inline tests/test_iqm_client.py:1036]:
```suggestion
        match=re.escape(
            f'Your IQM Client version {client_version} was built for a different version of IQM Server. '
            f'You might encounter issues. For the best experience, consider using a version '
            f'of IQM Client that satisfies {min_version} <= iqm-client < {max_version}.'
        ),
```

File: tests/test_iqm_client.py [hunk 3]
```diff
@@ -1006,3 +1009,63 @@ def test_get_dynamic_quantum_architecture_throws_json_decode_error_if_received_n
...
+    if server_version_diff == 0:
+        IQMClient(base_url)
+        assert len(recwarn) == 0
+    else:
+        with pytest.warns(
+            UserWarning,
+            match=re.escape(
+                f'IQM Client version {client_version} is incompatible with the IQM server, '
+                f'which requires {min_version} <= iqm-client < {max_version}'
+            ),
```
Comment (manzanillo) [inline tests/test_iqm_client.py:1068]:
```suggestion
            match=re.escape(
                f'Your IQM Client version {client_version} was built for a different version of IQM Server. '
                f'You might encounter issues. For the best experience, consider using a version '
                f'of IQM Client that satisfies {min_version} <= iqm-client < {max_version}.'
            ),
```

**Review by manzanillo — COMMENTED** [review]
I propose to formulate the warning less drastically, because it could be equally likely that the change is not impacting the user. It is still a warning and not an error.
I proposed some (slightly longer unfortunately :)) text

**Pull request merged**